ui/ducks/send/send.js

```diff
@@ -0,0 +1,1442 @@
+          recipient.warning = null;
+        }
+      }
+    },
+    updateRecipientSearchMode: (state, action) => {
+      state.recipient.mode = action.payload;
+    },
+    resetSendState: () => initialState,
+    validateAmountField: (state) => {
+      switch (true) {
+        // set error to INSUFFICIENT_FUNDS_ERROR if the account balance is lower
+        // than the total price of the transaction inclusive of gas fees.
+        case state.asset.type === ASSET_TYPES.NATIVE &&
+          !isBalanceSufficient({
+            amount: state.amount.value,
+            balance: state.asset.balance,
+            gasTotal: state.gas.gasTotal ?? '0x0',
+          }):
+          state.amount.error = INSUFFICIENT_FUNDS_ERROR;
+          break;
+        // set error to INSUFFICIENT_FUNDS_ERROR if the token balance is lower
+        // than the amount of token the user is attempting to send.
+        case state.asset.type === ASSET_TYPES.TOKEN &&
+          !isTokenBalanceSufficient({
+            tokenBalance: state.asset.balance ?? '0x0',
+            amount: state.amount.value,
+            decimals: state.asset.details.decimals,
+          }):
+          state.amount.error = INSUFFICIENT_TOKENS_ERROR;
+          break;
+        // if the amount is negative, set error to NEGATIVE_ETH_ERROR
+        // TODO: change this to NEGATIVE_ERROR and remove the currency bias.
+        case conversionGreaterThan(
+          { value: 0, fromNumericBase: 'dec' },
+          { value: state.amount.value, fromNumericBase: 'hex' },
+        ):
+          state.amount.error = NEGATIVE_ETH_ERROR;
+          break;
+        // If none of the above are true, set error to null
+        default:
+          state.amount.error = null;
+      }
+    },
+    validateGasField: (state) => {
+      // Checks if the user has enough funds to cover the cost of gas, always
+      // uses the native currency and does not take into account the amount
+      // being sent. If the user has enough to cover cost of gas but not gas
+      // + amount then the error will be displayed on the amount field.
+      const insufficientFunds = !isBalanceSufficient({
+        amount: '0x0',
+        balance: state.account.balance,
+        gasTotal: state.gas.gasTotal ?? '0x0',
+      });
+
+      state.gas.error = insufficientFunds ? INSUFFICIENT_FUNDS_ERROR : null;
+    },
+    validateSendState: (state) => {
+      switch (true) {
+        // 1 + 2. State is invalid when either gas or amount fields have errors
+        // 3. State is invalid if asset type is a token and the token details
+        //  are unknown.
+        // 4. State is invalid if no recipient has been added
+        // 5. State is invalid if gasLimit is less than the minimumGasLimit
+        case Boolean(state.amount.error):
+        case Boolean(state.gas.error):
+        case state.asset.type === ASSET_TYPES.TOKEN &&
+          state.asset.details === null:
+        case state.stage === SEND_STAGES.ADD_RECIPIENT:
```

Comment:
Should this consider an uninitialized send state as "valid" as well? :thinking: The old logic this replaces also consideres the "gasIsLoading" error case that was replaced by tracking the initialization state.

Reply:
the `.pending` action from the async thunk sets the state as invalid, this method is then called on a `.fulfilled` action from the async thunk. So in short, yes it should be and is considered invalid.